### PR TITLE
Docker-related changes for a local Runhouse server-in-Docker setup.

### PR DESCRIPTION
**Pull request by DenisYay**
This PR introduces Dockerfiles for the Debian/Ubuntu (`/docker/slim/Dockerfile`) and CUDA (`/docker/cuda/Dockerfile`) use cases.

It adds a new unit test in `test_http_server.py` that:

- Builds a local Docker instance encapsulating a `runhouse start` run
- Runs the local Docker instance
- Verifies the cluster is up by calling `check_server()` and confirming `is_up()` is True
- Tears down the Docker instance

The Dockerfile is tested using:
```
docker build --pull --rm -f "Dockerfile" -t runhouse:start "."
docker run --rm --shm-size=3gb -it -p 50052:50052 -p 6379:6379 -p 52365:52365 -p 22:22 runhouse:start
```

> Executing `/usr/local/bin/python -m runhouse.servers.http.http_server --host 0.0.0.0`
> INFO:     Started server process [26]
> INFO:     Waiting for application startup.
> INFO:     Application startup complete.
> INFO:     Uvicorn running on http://0.0.0.0:50052 (Press CTRL+C to quit)

### CODE REVIEW

File: dockerize.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Prerequisites: Start a local Docker image with Runhouse installed using the following commands:
```
Comment (jlewitt1) [inline dockerize.py:1]: Is this meant to be run as a test once the docker container is up? Maybe we should move this out of the main directory and into the tests file?

Comment (DenisYay) [reply dockerize.py]: Sounds good, done

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+FROM python:3.9.15
```
Comment (jlewitt1) [inline Dockerfile:1]: can we delete this from the root dir if we also save in the docker docker/base dir?

File: runhouse/docker/base/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+FROM python:3.9.15
```
Comment (jlewitt1) [inline runhouse/docker/base/Dockerfile:1]: worth using a smaller base like slim or alpine?

File: tests/conftest.py [hunk 4]
```diff
@@ -577,3 +577,50 @@ def create_gcs_bucket(bucket_name: str):
...
+
+    if result.returncode != 0:
+        print(result.stdout)
+
+    # Check for success
+    assert result.returncode == 0
+
+
+@pytest.fixture
+def local_docker():
```
Comment (jlewitt1) [inline tests/conftest.py:597]: can we make use of this fixture as part of a unittest somewhere? maybe incorporate with some of the checks made in `dockerize.py`?

File: runhouse/docker/base/Dockerfile (hunk 3 above)
Comment (DenisYay) [reply runhouse/docker/base/Dockerfile:1]: Updated

File: tests/dockerize.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+print("Runhouse version: ", rh.__version__)
+rh.configs.disable_data_collection()  # Workaround until we remove the usage of GCSClient from our code
+port = 50052
+c = rh.cluster(
+    name="local-docker", host=f"localhost:{port}", ssh_creds={"ssh_user": "root"}
+)
+print("Cluster created")
+c.up_if_not()
+print("Cluster is up: ", c.is_up())
```
Comment (carolineechen) [inline tests/dockerize.py:18]: can we turn this into a proper unittest and use asserts to check for expected values?

File: tests/conftest.py (hunk 4 above)
Comment (DenisYay) [reply tests/conftest.py:636]: Done

File: Dockerfile (hunk 2 above)
Comment (jlewitt1) [reply Dockerfile:1]: Gotcha yea I think renaming for now makes sense

File: tests/conftest.py (hunk 4 above)
Comment (DenisYay) [reply tests/conftest.py]: This is now used in: https://github.com/run-house/runhouse/blob/b1ae55ed21c9049c9c5a7530d2c556dfe7b3883a/tests/test_http_server.py

File: tests/dockerize.py (hunk 5 above)
Comment (DenisYay) [reply tests/dockerize.py:18]: Done in https://github.com/run-house/runhouse/blob/b1ae55ed21c9049c9c5a7530d2c556dfe7b3883a/tests/test_http_server.py

File: tests/conftest.py [hunk 6]
```diff
@@ -577,3 +578,102 @@ def create_gcs_bucket(bucket_name: str):
...
+    # Check for success
+    assert result.returncode == 0
+
+
+def run_shell_command(subprocess, cmd: list[str]):
+    # Run the command and wait for it to complete
+    result = subprocess.run(cmd, capture_output=True, text=True)
+
+    if result.returncode != 0:
+        print(result.stdout)
```
Comment (jlewitt1) [inline tests/conftest.py:602]: for more visibility if there's a failure maybe print stderr instead or in addition?

**Review by jlewitt1 — APPROVED** [review]

File: tests/conftest.py (hunk 6 above)
Comment (DenisYay) [reply tests/conftest.py:602]: Indeed, good idea, thanks!
Done.

File: tests/conftest.py [hunk 7]
```diff
@@ -577,3 +577,50 @@ def create_gcs_bucket(bucket_name: str):
...
+    import subprocess
+
+    # Requirement: need to install Docker on your machine
+
+    # Build the Docker image
+    run_shell_command(
+        subprocess,
+        [
+            "docker build --pull --rm -f '../Dockerfile' "
+            "--build-arg DOCKER_USER_PASSWORD_FILE=docker_user_passwd "
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply tests/conftest.py]: Added a check in the Docker file:

# Check if the password file exists, if not throw an error
RUN if [ ! -f "$DOCKER_USER_PASSWORD_FILE" ]; then echo "Error: Password file ${DOCKER_USER_PASSWORD_FILE} does not exist. Please create it and make sure it contains a password to be used for ssh access into the Docker container." && exit 1; fi

File: runhouse/docker/httpserver/Dockerfile
Comment (dongreenberg) [inline runhouse/docker/httpserver/Dockerfile:1]: Lets rename this "testing" so users don't accidentally use it

File: tests/conftest.py [hunk 8]
```diff
@@ -577,3 +578,104 @@ def create_gcs_bucket(bucket_name: str):
...
+            "runhouse:start",
+        ],
+    )
+
+    # Runhouse commands can now be run locally
+    rh.configs.disable_data_collection()  # Workaround until we remove the usage of GCSClient from our code
+    c = rh.cluster(
+        name="local-docker-slim",
+        host="localhost:50052",
+        ssh_creds={"ssh_user": "root"},
```
Comment (dongreenberg) [inline tests/conftest.py:669]: I thought we were setting up the password so we don't need to run as root?

**Review by dongreenberg — APPROVED** [review]
In a subsequent PR, fixtures which use key pairs in addition to the password ones would be great, and fleshing out the tests (as discussed live).

**Comment by DenisYay** [comment]
Re 

> In a subsequent PR, fixtures which use key pairs in addition to the password ones would be great, and fleshing out the tests (as discussed live).

Sounds good, thanks!

File: tests/conftest.py (hunk 8 above)
Comment (DenisYay) [reply tests/conftest.py:669]: Updated, thanks!

File: tests/conftest.py [hunk 9]
```diff
@@ -577,3 +578,104 @@ def create_gcs_bucket(bucket_name: str):
...
+    )
+    c.check_server()
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply tests/conftest.py:671]: Done

File: runhouse/docker/httpserver/Dockerfile
Comment (DenisYay) [reply runhouse/docker/httpserver/Dockerfile:1]: Done

File: runhouse/docker/cuda/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    echo "command=/usr/sbin/sshd -D" >> /etc/supervisor/conf.d/supervisord.conf && \
+    echo "stdout_logfile=/var/log/sshd.log" >> /etc/supervisor/conf.d/supervisord.conf && \
+    echo "stderr_logfile=/var/log/sshd.err" >> /etc/supervisor/conf.d/supervisord.conf && \
+    echo "[program:runhouse]" >> /etc/supervisor/conf.d/supervisord.conf && \
+    echo "command=runhouse start --host "0.0.0.0"" >> /etc/supervisor/conf.d/supervisord.conf && \
+    echo "stdout_logfile=/var/log/runhouse.log" >> /etc/supervisor/conf.d/supervisord.conf && \
+    echo "stderr_logfile=/var/log/runhouse.err" >> /etc/supervisor/conf.d/supervisord.conf
+
+# Runhouse server port
+EXPOSE 50052
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply runhouse/docker/cuda/Dockerfile:47]: Great point, added in all Dockerfiles.

File: .gitignore [hunk 11]
```diff
@@ -133,3 +133,11 @@ dmypy.json
 # IDE project files
 .idea/
 .vscode/
+
+# Config files
+rh/local_function/config.json
+rh/local_conda_env/config.json
+rh/tests/bert_ft/config.json
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply .gitignore:140]: Done.

**Pull request merged**